### PR TITLE
Squeeze is not reduction

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Adding a new `SqueezeOp` which is the opposite of `BroadcastOp`. For this PR, it just added this op with a few tests and uses this op to implement `squeeze`. This implementation of squeeze should be much cleaner than trivial reduction, and the goal is to replace the concept of trivial reduction with squeeze in the next few PRs. In the future, there will be no trivial reduction any more. All reductions will be non-trivial.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 1]
```diff
@@ -648,6 +648,88 @@ bool BroadcastOp::sameAs(const Statement* other) const {
...
+      "Cannot squeeze a non-tensor object.");
+
+  addOutput(out);
+  addInput(in);
+
+  if (!out->isA<TensorView>() || !in->isA<TensorView>()) {
+    return;
+  }
+
+  passkey.ir_container_->registerExpr(exprPasskey(), this);
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: This is copy-pasted from `BroadcastOp`, but I don't know why it is needed.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 2]
```diff
@@ -648,6 +648,88 @@ bool BroadcastOp::sameAs(const Statement* other) const {
...
+    Val* out,
+    Val* in,
+    std::vector<bool> is_squeeze_dims)
+    : Expr(passkey, ExprType::SqueezeOp),
+      out_(out),
+      in_(in),
+      is_squeeze_dims_(std::move(is_squeeze_dims)) {
+  // clang-tidy complains about out_ that it may be null.
+  TORCH_INTERNAL_ASSERT(out_ != nullptr);
+  TORCH_INTERNAL_ASSERT(in_ != nullptr);
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: Is this still the case? This is copied from BroadcastOp, and it looks weird...

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: Let's remove it if clang-tidy doesn't complain anymore.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: I don't think it's necessary. It should be done in the ctor call to `Expr`.

**Comment by naoyam** [comment]
So, if a fusion has a reduction of a size-1 domain, are we going to convert its expr to `SqueezeOp`? What if a tensor has non-size-1 domains and size-1 domains? Trivial reduction is a property associated with `IterDomain`, but `Squeeze` is an expression, so wouldn't there be some semantic mismatch?

**Comment by zasdfgbnm** [comment]
> So, if a fusion has a reduction of a size-1 domain, are we going to convert its expr to `SqueezeOp`? What if a tensor has non-size-1 domains and size-1 domains? Trivial reduction is a property associated with `IterDomain`, but `Squeeze` is an expression, so wouldn't there be some semantic mismatch?

We may or may not want to do so, but whichever way we choose, there shouldn't be any technical difficulty there. If we want to model size 1 reduction as squeeze, for the case where there are both trivial and non-trivial reductions, we can just make reduction a composite op as `input->squeeze->reduction->output`.

But actually, I prefer not to convert size 1 reduction as a squeeze. I prefer to handle it like a non-trivial reduction. Because, in PyTorch eager mode, the semantics of `squeeze` and `reduction` are not connected at all. For example, the command below shows that eager mode launches a reduction kernel on trivial reduction:
```
nsys nvprof python -c "import torch; torch.empty(1000, 1, device='cuda').sum(1)"
```

```
 Time (%)  Total Time (ns)  Instances  Avg (ns)  Med (ns)  Min (ns)  Max (ns)  StdDev (ns)     GridXYZ         BlockXYZ                                                     Name                                                
 --------  ---------------  ---------  --------  --------  --------  --------  -----------  --------------  --------------  ----------------------------------------------------------------------------------------------------
    100.0            4,512          1   4,512.0   4,512.0     4,512     4,512          0.0     2    1    1   128    1    1  void at::native::reduce_kernel<(int)128, (int)4, at::native::ReduceOp<float, at::native::func_wrapp…
```
And I will never expect a PyTorch user to use a `sum` to remove a size 1 dimension. If this happens, then it must be a dynamic shape that happens to be 1. But I will of course change many parts in nvfuser that are using reduction to do a squeeze.

If I just handle size 1 reduction as a non-trivial reduction, will this cause any trouble for us?

**Comment by naoyam** [comment]
The reason we handle size-1 reductions as trivial reductions is that we can ignore some of the constraints for reductions such as inlining, merge, etc. We also forward trivial trivial reductions when replaying expressions. Maybe if we use squeeze instead, we would not need to have those special handling of size-1 reductions, but it would just mean we would need to do something similar for squeeze. Maybe that's cleaner than trivial reductions? If that's the case, I think this is going to be a great improvement.

My feeling is that the trivial reduction has many similarities as the broadcast, and as we are finding that we would need some fundamental work for handling broadcasts in a more generic and robust way, I wouldn't be surprised if trivial reductions would also need similar work. So, if not urgently required, it seems to me that fixing broadcast first, and then revisiting size-1 reductions seems to make more sense.

**Comment by csarofeen** [comment]
> My feeling is that the trivial reduction has many similarities as the broadcast

I think all of the special handling is because of broadcast resolution which trivial reduction by definition doesn't have. Special handling we could need is the `ALMOSTEXACT` map in compute at and ID traversals, which would map domains that only differ by a merged broadcast dimension as mapped. I don't think we'd need anywhere near as much special handling for trivial reductions as we do for broadcast, but yes we might need a bit of special handling still.

**Comment by zasdfgbnm** [comment]
I think the biggest difference between squeeze and trivial reduction is that squeeze removes a broadcast dim immediately, but trivial reduction delays the removal of such dims by turning them into a reduction and then removing it in the next consumer. I kind of feel that the former is strictly cleaner. For non-trivial reduction, we are marking that dimension as reduction, because that would help us analyze and generate code. But for squeeze, I don't see any benefit from this extra step because trivial reduction is something that we just want to ignore and not generate code for it, and by adding this extra step, we need to add a lot of special handling for `if (id->isTrivialReduction())`, and this design would become a source of bugs. We already have a lot of special handling, and one less special handling would be nice because we are simply not smart enough to find all special cases before someone reports a bug (#2071, #2086). For the cases where we don't like the "constraints for reductions" on broadcast dims, then we just squeeze it and get rid of it before doing the reduction. For cases where we don't mind having these constraints, then simply treat it as a non-trivial reduction.

**Comment by naoyam** [comment]
Yeah, I'm not saying we shouldn't do this, and I see that squeeze would be a cleaner mechanism to remove broadcast domains. I'm just concerned if there could be redundant work if we did this now given that we would need to do something likely significant for broadcast domains. In some cases, we would need to replicate some of the logic for trivial reductions (looks like you already did this for BestEffortRepaly). But since it looks like you have already done a significant part of the change and I don't have any concrete concern, let's take this approach now.

Can all the use of trivial reductions be replaced with squeeze now? What would the follow-up PRs look like?

**Comment by zasdfgbnm** [comment]
> looks like you already did this for BestEffortRepaly

For BestEffortRepaly, I actually just copy-pasted `ConsumerForwardingInfo` into `ProducerForwardingInfo` with a few symmetric changes. In the future, when trivial reduction is totally gone, we can merge them into a single class.

> Can all the use of trivial reductions be replaced with squeeze now?

This is what I am hoping, but as I migrate more trivial reductions into squeeze, I wouldn't be surprised if I need to make a few more changes.

> What would the follow-up PRs look like?

I would add `TORCH_CHECK(no size 1 reductions);` to `lower2device.cpp` and run all tests to find all places that create size 1 reductions, and replace all of them with squeeze. When there is no more size 1 reductions, I will start to cleanup codes for trivial reductions.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: Oh, I see why it is needed. If we didn't register this manually, then the registration will happen after the constructor has returned in `IrBuilder::build`. But here we are using `PairwiseRootDomainMap` to validate the inputs and outputs, and `PairwiseRootDomainMap` requires the registration to be finished. I don't think this makes sense because it makes our codebase less modular. So I changed it to remove the usage of `PairwiseRootDomainMap`. Now input and output domains are validated manually inside the constructor. Doing so, I can now remove this spacial handling.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: I have removed this. Let's see if the linter will allow me to do so.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 3]
```diff
@@ -648,6 +648,88 @@ bool BroadcastOp::sameAs(const Statement* other) const {
...
+  }
+
+  passkey.ir_container_->registerExpr(exprPasskey(), this);
+
+  auto in_tv = in->as<TensorView>();
+  auto out_tv = out->as<TensorView>();
+  auto& in_dom = in_tv->domain()->noReductions();
+  TORCH_INTERNAL_ASSERT(
+      is_squeeze_dims_.size() == in_dom.size(),
+      "The dimensions of input tensor and does not math with is_squeeze_dims");
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: typo: match

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: fixed

**Comment by csarofeen** [comment]
I think this should be fine, and we actually just need to be come more tolerant to broadcast (when they're not actively being concretized). Which I think needs to be done by using almost exact maps, instead of exact maps. Broadcast is only significant in code generation for when we grab the stride entry from the meta data (making sure we don't grab the broadcasted dim's stride). When broadcasts are merged in, or iter domains are split by 1, it really doesn't effect analysis, or indexing. So we should become more tolerant to them showing up.

**Comment by zasdfgbnm** [comment]
> I think this should be fine, and we actually just need to be come more tolerant to broadcast (when they're not actively being concretized). Which I think needs to be done by using almost exact maps, instead of exact maps. Broadcast is only significant in code generation for when we grab the stride entry from the meta data (making sure we don't grab the broadcasted dim's stride). When broadcasts are merged in, or iter domains are split by 1, it really doesn't effect analysis, or indexing. So we should become more tolerant to them showing up.

If we split a broadcast dim by 5 to get an extent 5 broadcast dim, then merge with a concrete 5 to get a extent 25, what should the behavior of the almost-exact map for this case?

**Comment by csarofeen** [comment]
I'm not sure I have a good answer here (at least without generation of new IterDomains).
The property I would like in almost exact map is if two IterDomains should map in the almost exact map if:
- They can both use the same index if they're on the same memory space.

Really what I'm looking for is a common indexing map so we can use that to traverse. This gets a bit tricky for:
(1) Shift
(2) Gather
(3) Merged broadcasts

(3) Is the hard part to get right because if an expanded broadcast domain is used in view (so it gets merged) then it needs to be realized immediately i.e. `[b0{5}*i1{s1}]` results in a consumer IterDomain of `i2{5*s1}`. The problem here is if we ever have a resolved broadcast otherwise we expect there's some mapping where we have `i0{5}*i1{s1}` so we know how to index into `i1` instead of assuming `[b0{5}*i1{s1}]` has the same index of `[i1{s1}]`.

Today our references are selected opportunistically. If we have for example: 
`T0[b0, i1]`, `T1[b0, b1]`, and `T2[i0, b1]` and we do something like:
```
T3[b0, i1] = T0+T1
T4[i0, b1] = T1+T2
T4->merge(0)
```
Then inline everything into T4, we don't have a correct reference to the problem. We should have an `IterDomain` that's `I0*I1`, but we just have `b0*i1`, `b0*b1`, and `i0*i1`. So really what we should be doing for compute at map is generating the for loop domain, which doesn't have to even be in any actual reference. Then make sure we have almost exact map paths from every tensors loop domains into all the rfactor domains of all the tensors. Those would be used to index on, and we would identify reuse by using them.

Loop domains would be computed and would be disconnected from the direct tensor view transformations. Not completely as all the loop domains used for a TensorView would have to resolve the almost exact paths to its rfactor.

So back to the original question. I believe if an expanded broadcast is merged with an `IterDomain` in an `rFactor` transformation, we would need to realize it in the almost exact map, but if it's not then we wouldn't, as we would promote the loop `IterDomains` of the tensor view to contain as many realized broadcasts as the indexing pass would need.

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 4]
```diff
@@ -1160,6 +1160,8 @@ TensorView* reductionOp(
     TensorView* tv,
     bool keep_dim /*=false*/,
     DataType dtype /* DataType::Null */) {
+  // TODO: should we use squeeze for size 1 broadcast dim?
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/arith.cpp:1163]: Yeah, I think this would make sense.

File: torch/csrc/jit/codegen/cuda/lower_fusion_simplifier.cpp [hunk 5]
```diff
@@ -105,6 +105,14 @@ class UnaryOpInserter : private kir::ExprMutator {
         top, IrBuilder::create<UnaryOp>(container, UnaryOpType::Set, out, in));
   }
 
+  void handle(SqueezeOp* top) final {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_fusion_simplifier.cpp]: nit `top`->`sop`

File: torch/csrc/jit/codegen/cuda/mutator.cpp [hunk 6]
```diff
@@ -421,6 +421,20 @@ void OptOutMutator::mutate(BroadcastOp* bop) {
   IrBuilder::create<BroadcastOp>(container, out, in, flags);
 }
 
+void OptOutMutator::mutate(SqueezeOp* bop) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/mutator.cpp]: nit: `bop`->`sop`

File: torch/csrc/jit/codegen/cuda/ops/alias.cpp [hunk 7]
```diff
@@ -140,6 +141,44 @@ TensorView* flatten(TensorView* x, int64_t start_dim, int64_t end_dim) {
   return out;
 }
 
+TensorView* squeeze(TensorView* x, const std::vector<bool>& to_squeeze) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ops/alias.cpp:144]: Do you like the bool vector for squeeze over list of ints? Seems this is why you put a TODO to replace  the trivial reductions in view with squeeze? Or were you just worried about blanket enabling squeeze at this point?

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 8]
```diff
@@ -84,6 +84,11 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseRootDomainMap::map(
     broadcast_flags = bop->getBroadcastDimFlags();
   }
 
+  std::vector<bool> squeeze_flags;
+  if (SqueezeOp* bop = dynamic_cast<SqueezeOp*>(consumer_tv_->definition())) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: nit: `bop`->`sop`

**Review by csarofeen — COMMENTED** [review]
PR Looks good to me, but @naoyam can you review `root_domain_map.cpp` changes? I want to make sure we pick up non-mappable dims correctly with squeeze. If we have two paths where one path a broadcast gets squeezed and the other it's broadcasted in a normalization pattern, I want to be sure we'd process it correctly.

**Comment by naoyam** [comment]
I'll review this today.

File: torch/csrc/jit/codegen/cuda/ops/alias.cpp (hunk 7 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ops/alias.cpp:144]: > Do you like the bool vector for squeeze over list of ints? Seems this is why you put a TODO to replace the trivial reductions in view with squeeze? Or were you just worried about blanket enabling squeeze at this point?

`BroadcastOp` uses a bool vector, and `SqueezeOp` is symmetric with `BroadcastOp`, so I made them similar. The reason why I put a TODO to reduction, is mainly because I want first to find all places that should be directly using squeeze but is using reduction. After that, I will start to change the reduction to do squeeze.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 8 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/mutator.cpp (hunk 6 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/mutator.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 9]
```diff
@@ -580,46 +576,39 @@ BroadcastOp::BroadcastOp(
...
-    if (root_p2c.find(id) == root_p2c.end()) {
+  auto in_tv = in->as<TensorView>();
+  auto out_tv = out->as<TensorView>();
+  auto in_dom = TensorDomain::noReductions(in_tv->getMaybeRFactorDomain());
+  auto& out_dom = out_tv->getRootDomain();
+  TORCH_INTERNAL_ASSERT(
+      is_broadcast_dims_.size() == out_dom.size(),
+      "The dimensions of output tensor and does not match with is_broadcast_dims");
+
+  auto in_size = is_broadcast_dims_.size();
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: nit: it was a little difficult to follow here. Probably because at this point `is_broadcast_dims_.size()` isn't the size of input tensor. Also, `in_size--` when `is_broadcast_dims_[i]` is true was a little confusing. I'd say have `int num_new_broadcasts = 0;` here and increment it instead of `in_size--` would make the code easier to follow. The final assertion could be something like `num_new_broadcasts + in_dom.size() == is_braodcast_dims_.size()`.

**Review by naoyam — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/lower_fusion_simplifier.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/lower_fusion_simplifier.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 9 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: fixed
